File: mobile/src/features/finance/api/transactions.ts
import { apiFetch } from "@/lib/api/client";
import type {
  ApiResponse,
  PaginatedList,
  Transaction,
  TransactionCreate,
  TransactionUpdate,
} from "@/features/finance/types";

export interface ListTransactionsParams {
  dateFrom?: string;
  dateTo?: string;
  search?: string;
  status?: "draft" | "confirmed";
  limit?: number;
  offset?: number;
}

export function listTransactions(params?: ListTransactionsParams): Promise<PaginatedList<Transaction>> {
  const qs = new URLSearchParams();
  if (params?.dateFrom) qs.set('date_from', params.dateFrom);
  if (params?.dateTo) qs.set('date_to', params.dateTo);
  if (params?.search) qs.set('search', params.search);
  if (params?.status) qs.set('status', params.status);
  if (params?.limit != null) qs.set('limit', String(params.limit));
  if (params?.offset != null) qs.set('offset', String(params.offset));
  const query = qs.toString();
  return apiFetch<ApiResponse<Transaction[]>>(`/api/v1/transactions${query ? `?${query}` : ''}`)
    .then(r => ({ items: r.data, total: r.meta?.total ?? 0 }));
}

export function getTransaction(id: string): Promise<Transaction> {
  return apiFetch<ApiResponse<Transaction>>(`/api/v1/transactions/${id}`)
    .then(r => r.data);
}

export function createTransaction(data: TransactionCreate): Promise<Transaction> {
  return apiFetch<ApiResponse<Transaction>>("/api/v1/transactions", {
    method: "POST",
    body: JSON.stringify(data),
  }).then(r => r.data);
}

export function updateTransaction(id: string, data: TransactionUpdate): Promise<Transaction> {
  return apiFetch<ApiResponse<Transaction>>(`/api/v1/transactions/${id}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  }).then(r => r.data);
}

export function deleteTransaction(id: string): Promise<void> {
  return apiFetch<void>(`/api/v1/transactions/${id}`, { method: "DELETE" });
}

File: backend/app/domains/finance/routers/transactions.py
import uuid
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Query

from app.core.auth import CurrentUser
from app.core.exceptions import ForbiddenError
from app.core.response import ApiResponse, ok, paginated
from app.domains.finance import service
from app.domains.finance.models import TransactionStatus
from app.domains.finance.routers.deps import DbSession
from app.domains.finance.schemas import TransactionCreate, TransactionRead, TransactionUpdate

router = APIRouter(tags=["Transactions"])


@router.get("/transactions", response_model=ApiResponse[list[TransactionRead]])
async def list_transactions(
    user_id: CurrentUser,
    session: DbSession,
    account_id: Annotated[uuid.UUID | None, Query()] = None,
    date_from: Annotated[date | None, Query()] = None,
    date_to: Annotated[date | None, Query()] = None,
    search: Annotated[str | None, Query(max_length=200)] = None,
    status: Annotated[TransactionStatus | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=1000)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> ApiResponse[list[TransactionRead]]:
    items, total = await service.list_transactions(
        session, user_id,
        account_id=account_id, date_from=date_from, date_to=date_to,
        search=search, status=status, limit=limit, offset=offset,
    )
    return paginated(items, total=total, limit=limit, offset=offset)


@router.post("/transactions", response_model=ApiResponse[TransactionRead], status_code=201)
async def create_transaction(
    user_id: CurrentUser, session: DbSession, data: TransactionCreate
) -> ApiResponse[TransactionRead]:
    item = await service.create_transaction(session, user_id, data)
    return ok(item, message="created")


@router.get("/transactions/{transaction_id}", response_model=ApiResponse[TransactionRead])
async def get_transaction(
    transaction_id: uuid.UUID, user_id: CurrentUser, session: DbSession
) -> ApiResponse[TransactionRead]:
    tx = await service.get_transaction_or_404(session, transaction_id)
    if tx.user_id != user_id:
        raise ForbiddenError("You don't own this transaction")
    return ok(tx)


@router.patch("/transactions/{transaction_id}", response_model=ApiResponse[TransactionRead])
async def update_transaction(
    transaction_id: uuid.UUID, user_id: CurrentUser, session: DbSession, data: TransactionUpdate
) -> ApiResponse[TransactionRead]:
    item = await service.update_transaction(session, user_id, transaction_id, data)
    return ok(item)


@router.delete("/transactions/{transaction_id}", status_code=204)
async def delete_transaction(
    transaction_id: uuid.UUID, user_id: CurrentUser, session: DbSession
) -> None:
    await service.delete_transaction(session, user_id, transaction_id)
